File: tools/wavefield2vtk_detail.py
#!/usr/bin/env python3
"""Convert all strain snapshot record files to VTK (mesh + GLL points).

Merges all MPI ranks, 3 source directions (x, y, z), and all time steps
into per-timestep VTK files. Includes hex cells + GLL-derived edge (LINE),
face (QUAD), and sub-volume (HEX) cells with per-point strain.

Usage:
    cd examples/halfspace/
    python ../../tools/wavefield2vtk_detail.py

Reads:
    mesh.h5
    wavefields/{x,y,z}/record_{r}.h5  (all ranks r)
    config.h5                       — for snapshot stride

Writes:
    vtk/wavefield_N.vtk              N = solver step number
"""

import glob
import os
import re

import h5py
import numpy as np

_HEX_FACES = [[0, 3, 2, 1], [4, 5, 6, 7], [0, 1, 5, 4], [3, 7, 6, 2], [0, 4, 7, 3], [1, 2, 6, 5]]

_VOIGT_LABELS = ["xx", "yy", "zz", "xy", "xz", "yz"]
_DIRECTIONS = ["x", "y", "z"]


def resolve_cell_vertices(cell_to_surface, surface_to_edge, edge_to_vertex, cell_idx):
    signed_surfaces = cell_to_surface[cell_idx]
    local_to_global = {}
    for fi in range(6):
        sid_signed = signed_surfaces[fi]
        sid = int(abs(sid_signed)) - 1
        canonical_edges = surface_to_edge[sid]
        if sid_signed > 0:
            signed_edges = canonical_edges
        else:
            signed_edges = [
                -canonical_edges[3],
                -canonical_edges[2],
                -canonical_edges[1],
                -canonical_edges[0],
            ]
        for k in range(4):
            eid = int(abs(signed_edges[k])) - 1
            gv1, gv2 = edge_to_vertex[eid]
            gv1 -= 1
            gv2 -= 1
            lvk = _HEX_FACES[fi][k]
            lvk_next = _HEX_FACES[fi][(k + 1) % 4]
            if signed_edges[k] > 0:
                local_to_global[lvk] = gv1
                local_to_global[lvk_next] = gv2
            else:
                local_to_global[lvk] = gv2
                local_to_global[lvk_next] = gv1
    return [local_to_global[lv] for lv in range(8)]


def build_global_connectivity(cell_to_surface, surface_to_edge, edge_to_vertex):
    n_cell = cell_to_surface.shape[0]
    connectivity = np.zeros((n_cell, 8), dtype=np.int64)
    for ci in range(n_cell):
        conn = resolve_cell_vertices(cell_to_surface, surface_to_edge, edge_to_vertex, ci)
        connectivity[ci] = conn
    return connectivity


# ── GLL topology helpers ────────────────────────────────────────────────


def _gll_idx(i, j, k, ngll):
    return i * ngll * ngll + j * ngll + k


def build_gll_cell_template(ngll):
    """Return (edge_lines, face_quads, sub_hexes) templates (local 0-based indices)."""

    def idx(i, j, k):
        return _gll_idx(i, j, k, ngll)

    edge_lines = []
    for j, k in [(0, 0), (ngll - 1, 0), (0, ngll - 1), (ngll - 1, ngll - 1)]:
        for i in range(ngll - 1):
            edge_lines.append((idx(i, j, k), idx(i + 1, j, k)))
    for i, k in [(0, 0), (ngll - 1, 0), (0, ngll - 1), (ngll - 1, ngll - 1)]:
        for j in range(ngll - 1):
            edge_lines.append((idx(i, j, k), idx(i, j + 1, k)))
    for i, j in [(0, 0), (ngll - 1, 0), (ngll - 1, ngll - 1), (0, ngll - 1)]:
        for k in range(ngll - 1):
            edge_lines.append((idx(i, j, k), idx(i, j, k + 1)))

    face_quads = []
    for i in range(ngll - 1):
        for j in range(ngll - 1):
            face_quads.append(
                (idx(i, j, 0), idx(i + 1, j, 0), idx(i + 1, j + 1, 0), idx(i, j + 1, 0))
            )
    for i in range(ngll - 1):
        for j in range(ngll - 1):
            face_quads.append(
                (
                    idx(i, j, ngll - 1),
                    idx(i + 1, j, ngll - 1),
                    idx(i + 1, j + 1, ngll - 1),
                    idx(i, j + 1, ngll - 1),
                )
            )
    for i in range(ngll - 1):
        for k in range(ngll - 1):
            face_quads.append(
                (idx(i, 0, k), idx(i + 1, 0, k), idx(i + 1, 0, k + 1), idx(i, 0, k + 1))
            )
    for i in range(ngll - 1):
        for k in range(ngll - 1):
            face_quads.append(
                (
                    idx(i, ngll - 1, k),
                    idx(i + 1, ngll - 1, k),
                    idx(i + 1, ngll - 1, k + 1),
                    idx(i, ngll - 1, k + 1),
                )
            )
    for j in range(ngll - 1):
        for k in range(ngll - 1):
            face_quads.append(
                (idx(0, j, k), idx(0, j + 1, k), idx(0, j + 1, k + 1), idx(0, j, k + 1))
            )
    for j in range(ngll - 1):
        for k in range(ngll - 1):
            face_quads.append(
                (
                    idx(ngll - 1, j, k),
                    idx(ngll - 1, j + 1, k),
                    idx(ngll - 1, j + 1, k + 1),
                    idx(ngll - 1, j, k + 1),
                )
            )

    sub_hexes = []
    for i in range(ngll - 1):
        for j in range(ngll - 1):
            for k in range(ngll - 1):
                sub_hexes.append(
                    (
                        idx(i, j, k),
                        idx(i + 1, j, k),
                        idx(i + 1, j + 1, k),
                        idx(i, j + 1, k),
                        idx(i, j, k + 1),
                        idx(i + 1, j, k + 1),
                        idx(i + 1, j + 1, k + 1),
                        idx(i, j + 1, k + 1),
                    )
                )

    return edge_lines, face_quads, sub_hexes


def build_all_gll_cells(edge_template, face_template, sub_template, n_cell, ngll, n_mesh_vert):
    """Build full GLL cell arrays + element map for all cells.

    Returns (edge_arr, face_arr, sub_arr, n_edge, n_face, n_sub, gll_elem_map).
    """
    gll_per_cell = ngll**3

    n_edge = n_cell * len(edge_template)
    n_face = n_cell * len(face_template)
    n_sub = n_cell * len(sub_template)

    edge_arr = np.zeros(n_edge * 3, dtype=np.int32)
    face_arr = np.zeros(n_face * 5, dtype=np.int32)
    sub_arr = np.zeros(n_sub * 9, dtype=np.int32)

    for e in range(n_cell):
        base = n_mesh_vert + e * gll_per_cell
        for li, (a, b) in enumerate(edge_template):
            pos = (e * len(edge_template) + li) * 3
            edge_arr[pos] = 2
            edge_arr[pos + 1] = base + a
            edge_arr[pos + 2] = base + b
        for li, (a, b, c, d) in enumerate(face_template):
            pos = (e * len(face_template) + li) * 5
            face_arr[pos] = 4
            face_arr[pos + 1] = base + a
            face_arr[pos + 2] = base + b
            face_arr[pos + 3] = base + c
            face_arr[pos + 4] = base + d
        for li, corners in enumerate(sub_template):
            pos = (e * len(sub_template) + li) * 9
            sub_arr[pos] = 8
            for ci, corner in enumerate(corners):
                sub_arr[pos + 1 + ci] = base + corner

    gll_elem_map = np.concatenate(
        [
            np.repeat(np.arange(n_cell, dtype=np.int32), len(edge_template)),
            np.repeat(np.arange(n_cell, dtype=np.int32), len(face_template)),
            np.repeat(np.arange(n_cell, dtype=np.int32), len(sub_template)),
        ]
    )
    return edge_arr, face_arr, sub_arr, n_edge, n_face, n_sub, gll_elem_map


# ── VTK writer (legacy binary) ──────────────────────────────────────────


def write_vtu(
    path,
    vertex_coords,
    connectivity,
    cell_fields,
    point_fields,
    n_mesh_vert,
    edge_arr=None,
    face_arr=None,
    sub_arr=None,
    n_edge=0,
    n_face=0,
    n_sub=0,
    gll_elem_map=None,
):
    """Write VTK with hex cells + GLL edge/face/sub cells (binary).

    Points = [mesh_vertices | GLL_points].
    Cell ordering: [mesh_hexes | edge_LINEs | face_QUADs | sub_HEXes]
    """
    n_vert = vertex_coords.shape[0]
    n_hex = connectivity.shape[0]
    has_detail = point_fields is not None
    if has_detail:
        total_cells = n_hex + n_edge + n_face + n_sub
        total_ints = n_hex * 9 + n_edge * 3 + n_face * 5 + n_sub * 9
    else:
        total_cells = n_hex
        total_ints = n_hex * 9

    with open(path, "wb") as f:
        f.write(b"# vtk DataFile Version 3.0\n")
        f.write(b"wavefield snapshot converted to VTK (detail)\n")
        f.write(b"BINARY\n")
        f.write(b"DATASET UNSTRUCTURED_GRID\n")

        f.write(f"POINTS {n_vert} float\n".encode())
        f.write(np.ascontiguousarray(vertex_coords, dtype=">f4").tobytes())
        f.write(b"\n")

        f.write(f"CELLS {total_cells} {total_ints}\n".encode())
        hex_arr = np.zeros(n_hex * 9, dtype=np.int32)
        hex_arr[0::9] = 8
        for i in range(8):
            hex_arr[1 + i :: 9] = connectivity[:, i].astype(np.int32)
        if has_detail:
            cell_arr = np.concatenate([hex_arr, edge_arr, face_arr, sub_arr])
        else:
            cell_arr = hex_arr
        f.write(np.ascontiguousarray(cell_arr, dtype=">i4").tobytes())
        f.write(b"\n")

        f.write(f"CELL_TYPES {total_cells}\n".encode())
        if has_detail:
            types_arr = np.concatenate(
                [
                    np.full(n_hex, 12, dtype=np.int32),
                    np.full(n_edge, 3, dtype=np.int32),
                    np.full(n_face, 9, dtype=np.int32),
                    np.full(n_sub, 12, dtype=np.int32),
                ]
            )
        else:
            types_arr = np.full(n_hex, 12, dtype=np.int32)
        f.write(np.ascontiguousarray(types_arr, dtype=">i4").tobytes())
        f.write(b"\n")

        f.write(f"CELL_DATA {total_cells}\n".encode())
        for name, data in cell_fields.items():
            data_padded = np.zeros(total_cells, dtype=data.dtype)
            data_padded[:n_hex] = data
            if gll_elem_map is not None:
                data_padded[n_hex:] = data[gll_elem_map]
            f.write(f"SCALARS {name} float 1\n".encode())
            f.write(b"LOOKUP_TABLE default\n")
            f.write(np.ascontiguousarray(data_padded, dtype=">f4").tobytes())
            f.write(b"\n")

        if has_detail:
            f.write(f"POINT_DATA {n_vert}\n".encode())
            for name, data in point_fields.items():
                f.write(f"SCALARS {name} float 1\n".encode())
                f.write(b"LOOKUP_TABLE default\n")
                f.write(np.ascontiguousarray(data, dtype=">f4").tobytes())
                f.write(b"\n")


# ── Record file discovery ──────────────────────────────────────────────


def find_record_files(wave_dir):
    pattern = os.path.join(wave_dir, "record_*.h5")
    files = glob.glob(pattern)
    files.sort(key=lambda p: int(re.search(r"record_(\d+)\.h5$", p).group(1)))
    return files


# ── Main ───────────────────────────────────────────────────────────────


def main():
    cwd = os.getcwd()
    mesh_path = os.path.join(cwd, "mesh.h5")
    config_path = os.path.join(cwd, "config.h5")

    print(f"[wavefield2vtk_detail] Reading {mesh_path}")
    with h5py.File(mesh_path, "r") as f:
        topo = f["topology"]
        vertex_to_coord = topo["vertex_to_coord"][:]
        edge_to_vertex = topo["edge_to_vertex"][:]
        surface_to_edge = topo["surface_to_edge"][:]
        cell_to_surface = topo["cell_to_surface"][:]
        is_pml = np.zeros(cell_to_surface.shape[0], dtype=np.int8)
        if "field/element/is_pml" in f:
            is_pml[:] = f["field/element/is_pml"][:]
        gll_coords = f["field/element/coords"][:]
    n_cell = cell_to_surface.shape[0]
    n_mesh_vert = vertex_to_coord.shape[0]
    ngll = gll_coords.shape[1]
    print(f"  Global cells: {n_cell}, vertices: {n_mesh_vert}, NGLL: {ngll}")

    print("[wavefield2vtk_detail] Resolving hexahedral connectivity...")
    connectivity = build_global_connectivity(cell_to_surface, surface_to_edge, edge_to_vertex)

    # ── Build GLL point array (reused for all timesteps) ──
    print("[wavefield2vtk_detail] Building GLL point array...")
    gll_pt_list = []
    for ci in range(n_cell):
        gll_pt_list.append(gll_coords[ci].reshape(-1, 3))
    gll_points = np.concatenate(gll_pt_list, axis=0) if gll_pt_list else np.empty((0, 3))
    all_points = np.concatenate([vertex_to_coord, gll_points], axis=0)
    gll_per_cell = ngll**3
    total_gll = n_cell * gll_per_cell
    print(f"  GLL points per cell: {gll_per_cell}, total GLL: {total_gll}")
    print(f"  Total points: {all_points.shape[0]}")

    # ── Build GLL topology cells (reused for all timesteps) ──
    print("[wavefield2vtk_detail] Building GLL topology...")
    edge_tmpl, face_tmpl, sub_tmpl = build_gll_cell_template(ngll)
    edge_arr, face_arr, sub_arr, n_edge, n_face, n_sub, gll_elem_map = build_all_gll_cells(
        edge_tmpl, face_tmpl, sub_tmpl, n_cell, ngll, n_mesh_vert
    )
    print(
        f"  GLL topology: {n_edge}L + {n_face}Q + {n_sub}H = {n_edge + n_face + n_sub} GLL cells"
    )

    # ── Find record files per direction ──
    record_paths = {}
    for d in _DIRECTIONS:
        wave_dir = os.path.join(cwd, f"wavefields/{d}")
        files = find_record_files(wave_dir)
        if not files:
            print(f"[wavefield2vtk_detail] Error: no record_*.h5 files in {wave_dir}")
            return 1
        record_paths[d] = files
        print(f"  wavefields/{d}/: {len(files)} rank files")

    # ── Read metadata ──
    with h5py.File(record_paths["x"][0], "r") as f:
        n_snapshots = f["strain"].shape[0]
    print(f"  Snapshots: {n_snapshots}")

    stride = 1
    if os.path.isfile(config_path):
        try:
            with h5py.File(config_path, "r") as f:
                stride = int(f["simulation"].attrs.get("snapshot_stride", 1))
        except Exception:
            pass
    print(f"  Snapshot stride: {stride}")

    # ── Pre-read local_element_ids ──
    local_eids_list = []
    for path in record_paths["x"]:
        with h5py.File(path, "r") as f:
            local_eids_list.append(f["local_element_ids"][:].copy())
    for d in ("y", "z"):
        for ri, path in enumerate(record_paths[d]):
            with h5py.File(path, "r") as f:
                eids = f["local_element_ids"][:]
                if not np.array_equal(eids, local_eids_list[ri]):
                    print(f"[wavefield2vtk_detail] Error: element ID mismatch in {path}")
                    return 1

    # ── Open all record files ──
    files = {}
    for d in _DIRECTIONS:
        files[d] = [h5py.File(p, "r") for p in record_paths[d]]

    cell_fields_pml = {"PML_flag": is_pml.astype(np.float64)}

    out_dir = os.path.join(cwd, "vtk")
    os.makedirs(out_dir, exist_ok=True)

    # ── Iterate snapshots ──
    for snap_idx in range(n_snapshots):
        step_num = snap_idx * stride

        dir_strain = {}
        for d in _DIRECTIONS:
            gs = np.zeros((n_cell, gll_per_cell, 6), dtype=np.float64)
            for ri, f in enumerate(files[d]):
                snap = f["strain"][snap_idx]
                gs[local_eids_list[ri]] = snap.reshape(-1, gll_per_cell, 6)
            dir_strain[d] = gs

        # Build point data: 18 strain fields
        strain_field_names = [f"strain_{vl}_{d}" for d in _DIRECTIONS for vl in _VOIGT_LABELS]
        point_fields = {}
        for fi, name in enumerate(strain_field_names):
            di = fi // 6
            ci = fi % 6
            arr = np.zeros(all_points.shape[0], dtype=np.float64)
            arr[n_mesh_vert:] = dir_strain[_DIRECTIONS[di]][:, :, ci].ravel()
            point_fields[name] = arr

        out_path = os.path.join(out_dir, f"wavefield_{step_num}.vtk")
        print(f"[wavefield2vtk_detail] Writing {out_path}")
        write_vtu(
            out_path,
            all_points,
            connectivity,
            cell_fields_pml,
            point_fields,
            n_mesh_vert,
            edge_arr,
            face_arr,
            sub_arr,
            n_edge,
            n_face,
            n_sub,
            gll_elem_map,
        )

    for d in _DIRECTIONS:
        for f in files[d]:
            f.close()

    print(f"  Done. {n_snapshots} files written to {out_dir}/")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/cpp/wavefield2vtk_detail.cpp
// tools/cpp/wavefield2vtk_detail.cpp
// C++ accelerated wavefield2vtk_detail — reads vertex strain snapshots,
// writes per-step VTK with strain as point data at mesh vertices.
//
// Unlike gf_wavefield2vtk (cell-corner average), this writes the raw
// per-vertex strain directly as 18 POINT_DATA fields (6 Voigt × 3 directions).

#include <dirent.h>
#include <omp.h>
#include <sys/stat.h>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstring>
#include <iostream>
#include <regex>
#include <set>
#include <string>
#include <vector>

#include "h5io.hh"
#include "topology.hh"
#include "vtk_writer.hh"

using namespace gf_topology;
using namespace gf_h5io;

static const char* VOIGT_LABELS[6] = {"xx", "yy", "zz", "xy", "xz", "yz"};
static const char* DIRECTIONS[3] = {"x", "y", "z"};

static void print_usage(const char* prog) {
    std::fprintf(stderr,
                 "Usage: %s [options]\n"
                 "  --verbose, -v         Verbose output\n"
                 "  --model PATH          Model file (default: model.h5)\n"
                 "  --config PATH         Config file (default: config.h5)\n"
                 "  --vtk-dir DIR         Output directory (default: vtk)\n"
                 "  --snap N              Single snapshot index (default: all)\n"
                 "  --help, -h            This help\n",
                 prog);
}

struct RecordFile {
    std::string path;
    int rank;
    int step;  // solver step number
};

// ── Record file discovery ──────────────────────────────────────────────

/// Find per-step record files matching record_{rank}_{step_num}.h5.
static std::vector<RecordFile> find_record_files(const std::string& wave_dir, int step_num) {
    std::vector<RecordFile> files;
    DIR* dir = opendir(wave_dir.c_str());
    if (!dir)
        return files;
    struct dirent* entry;
    while ((entry = readdir(dir)) != nullptr) {
        std::string name(entry->d_name);
        std::regex re("record_(\\d+)_(\\d+)\\.h5$");
        std::smatch m;
        if (std::regex_search(name, m, re) && m.size() > 2) {
            int rank = std::stoi(m[1].str());
            int fstep = std::stoi(m[2].str());
            if (fstep == step_num) {
                files.push_back({wave_dir + "/" + name, rank, fstep});
            }
        }
    }
    closedir(dir);
    std::sort(files.begin(), files.end(),
              [](const RecordFile& a, const RecordFile& b) { return a.rank < b.rank; });
    return files;
}

/// Discover all unique solver step numbers from record files in a directory.
static std::vector<int> discover_steps(const std::string& wave_dir) {
    std::set<int> steps;
    DIR* dir = opendir(wave_dir.c_str());
    if (!dir)
        return {};
    struct dirent* entry;
    while ((entry = readdir(dir)) != nullptr) {
        std::string name(entry->d_name);
        std::regex re("record_(\\d+)_(\\d+)\\.h5$");
        std::smatch m;
        if (std::regex_search(name, m, re) && m.size() > 2) {
            steps.insert(std::stoi(m[2].str()));
        }
    }
    closedir(dir);
    std::vector<int> sorted(steps.begin(), steps.end());
    std::sort(sorted.begin(), sorted.end());
    return sorted;
}

// ── HDF5 readers ──────────────────────────────────────────────────────

/// Read vertex_ids dataset from a record file.
static std::vector<int64_t> read_vertex_ids(const std::string& path) {
    H5File f(path);
    return read_int64_1d(f.id(), "vertex_ids");
}

/// Read strain[0, :, :] as float32 from a per-step record file.
/// Returns flattened [n_local * 6] float values.
static std::vector<float> read_step_strain(const std::string& path, int64_t n_local) {
    hid_t fid = H5Fopen(path.c_str(), H5F_ACC_RDONLY, H5P_DEFAULT);
    if (fid < 0)
        throw std::runtime_error("Cannot open: " + path);
    hid_t dset = H5Dopen2(fid, "strain", H5P_DEFAULT);
    if (dset < 0) {
        H5Fclose(fid);
        throw std::runtime_error("No strain dataset in " + path);
    }
    // Per-step files have shape [1, n_local, 6]
    std::vector<float> buf(n_local * 6);
    hsize_t count[3] = {1, (hsize_t)n_local, 6};
    hid_t mspace = H5Screate_simple(3, count, nullptr);
    H5Dread(dset, H5T_NATIVE_FLOAT, mspace, H5S_ALL, H5P_DEFAULT, buf.data());
    H5Sclose(mspace);
    H5Dclose(dset);
    H5Fclose(fid);
    return buf;
}

/// Scatter per-rank strain into global vertex array.
/// strain_out: [n_vertex, 6] output (pre-allocated, caller zeros it).
static void scatter_strain(const std::vector<float>& rank_strain,
                           const std::vector<int64_t>& vertex_ids, int64_t n_vertex,
                           std::vector<float>& strain_out) {
    int64_t n_local = (int64_t)vertex_ids.size();
    for (int64_t li = 0; li < n_local; ++li) {
        int64_t gid = vertex_ids[li] - 1;  // 1-based → 0-based
        if (gid < 0 || gid >= n_vertex)
            continue;
        const float* src = rank_strain.data() + li * 6;
        float* dst = strain_out.data() + gid * 6;
        for (int c = 0; c < 6; ++c)
            dst[c] = src[c];
    }
}

// ── Main ──────────────────────────────────────────────────────────────

int main(int argc, char** argv) {
    bool verbose = false;
    std::string model_path = "model.h5";
    std::string config_path = "config.h5";
    std::string vtk_dir = "vtk";
    int snap_only = -1;  // -1 = process all snapshots

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--verbose" || arg == "-v")
            verbose = true;
        else if (arg == "--model" && i + 1 < argc)
            model_path = argv[++i];
        else if (arg == "--config" && i + 1 < argc)
            config_path = argv[++i];
        else if (arg == "--vtk-dir" && i + 1 < argc)
            vtk_dir = argv[++i];
        else if (arg == "--snap" && i + 1 < argc)
            snap_only = std::stoi(argv[++i]);
        else if (arg == "--help" || arg == "-h") {
            print_usage(argv[0]);
            return 0;
        }
    }

    // Create output directory
    struct stat st = {};
    if (stat(vtk_dir.c_str(), &st) != 0) {
        if (mkdir(vtk_dir.c_str(), 0755) != 0) {
            std::cerr << "Warning: could not create " << vtk_dir << "\n";
        }
    }

    // ── Read mesh topology ──────────────────────────────────────────
    if (verbose)
        std::cout << "[wavefield2vtk_detail] Reading " << model_path << "\n";
    H5File fm(model_path);
    hid_t topo_gid = H5Gopen2(fm.id(), "topology", H5P_DEFAULT);
    if (topo_gid < 0) {
        std::cerr << "Error: no /topology group\n";
        return 1;
    }
    auto vertex_to_coord = read_float64_2d(topo_gid, "vertex_to_coord");
    int64_t n_vert = (int64_t)vertex_to_coord.size();
    auto edge_to_vertex = read_int64_2d(topo_gid, "edge_to_vertex");
    auto surface_to_edge = read_int64_2d(topo_gid, "surface_to_edge");
    auto cell_to_surface = read_int64_2d(topo_gid, "cell_to_surface");
    H5Gclose(topo_gid);
    int64_t n_cell = (int64_t)cell_to_surface.size();

    // PML flag (optional)
    std::vector<int8_t> is_pml(n_cell, 0);
    if (dataset_exists(fm.id(), "field/element/is_pml")) {
        auto pml = read_int32_1d(fm.id(), "field/element/is_pml");
        for (size_t i = 0; i < pml.size() && i < (size_t)n_cell; ++i)
            is_pml[i] = (int8_t)pml[i];
    }
    if (verbose)
        std::cout << "  Global cells: " << n_cell << ", vertices: " << n_vert << "\n";

    // ── Resolve hex connectivity ─────────────────────────────────────
    if (verbose)
        std::cout << "[wavefield2vtk_detail] Resolving hex connectivity...\n";
    std::vector<int64_t> connectivity(n_cell * 8, -1);
    for (int64_t ci = 0; ci < n_cell; ++ci) {
        auto conn =
            resolve_cell_vertices(cell_to_surface[ci].data(), surface_to_edge, edge_to_vertex);
        for (int j = 0; j < 8; ++j)
            connectivity[ci * 8 + j] = conn[j];
    }

    // ── Read snapshot stride from config.h5 ──────────────────────────
    int stride = 1;
    try {
        H5File cf(config_path);
        stride = read_attr_int32_group(cf.id(), "simulation", "snapshot_stride");
    } catch (...) {
    }
    if (verbose)
        std::cout << "  Snapshot stride: " << stride << "\n";

    // ── Pre-compute vertex coords as float32 ─────────────────────────
    std::vector<float> vtx_coords_f32(n_vert * 3);
    for (int64_t i = 0; i < n_vert; ++i) {
        vtx_coords_f32[i * 3 + 0] = (float)vertex_to_coord[i][0];
        vtx_coords_f32[i * 3 + 1] = (float)vertex_to_coord[i][1];
        vtx_coords_f32[i * 3 + 2] = (float)vertex_to_coord[i][2];
    }

    // Strain field names: strain_{voigt}_{direction}
    std::vector<std::string> strain_field_names;
    for (int di = 0; di < 3; ++di)
        for (int vi = 0; vi < 6; ++vi)
            strain_field_names.push_back(std::string("strain_") + VOIGT_LABELS[vi] + "_" +
                                         DIRECTIONS[di]);

    // ── Discover available steps ─────────────────────────────────────
    std::vector<int> all_steps;
    if (snap_only >= 0) {
        all_steps.push_back(snap_only * stride);
    } else {
        all_steps = discover_steps("./wavefields/x/");
        if (all_steps.empty()) {
            std::cerr << "Error: no record files found in wavefields/x/\n";
            return 1;
        }
        if (verbose)
            std::cout << "  Found " << all_steps.size() << " snapshots\n";
    }
    int n_snapshots = (int)all_steps.size();

    // ── Pre-read vertex_ids from first step (validate across directions) ──
    std::vector<std::vector<int64_t>> vertex_id_list;
    {
        auto files = find_record_files("./wavefields/x/", all_steps[0]);
        if (files.empty()) {
            std::cerr << "Error: no record files for step " << all_steps[0] << "\n";
            return 1;
        }
        int n_ranks = (int)files.size();
        if (verbose)
            std::cout << "  Ranks: " << n_ranks << "\n";

        vertex_id_list.resize(n_ranks);
        for (int ri = 0; ri < n_ranks; ++ri) {
            vertex_id_list[ri] = read_vertex_ids(files[ri].path);
        }

        // Verify vertex_ids match across directions for rank 0
        for (int di = 1; di < 3; ++di) {
            std::string dwave = "./wavefields/" + std::string(DIRECTIONS[di]);
            auto dfiles = find_record_files(dwave, files[0].step);
            if (dfiles.empty()) {
                std::cerr << "Error: no files in " << dwave << "\n";
                return 1;
            }
            auto dvids = read_vertex_ids(dfiles[0].path);
            if (dvids != vertex_id_list[0]) {
                std::cerr << "Error: vertex_id mismatch in " << dwave << "\n";
                return 1;
            }
        }
        if (verbose)
            std::cout << "  Verified vertex IDs match across all directions\n";
    }

    // ── Process each snapshot ────────────────────────────────────────
    for (int si = 0; si < n_snapshots; ++si) {
        int step_num = all_steps[si];

        // Discover record files for this step across all directions
        std::vector<std::vector<RecordFile>> step_files(3);
        for (int di = 0; di < 3; ++di) {
            std::string wave_dir = "./wavefields/" + std::string(DIRECTIONS[di]);
            step_files[di] = find_record_files(wave_dir, step_num);
            if (step_files[di].empty()) {
                std::cerr << "Error: no record files for step " << step_num << " in " << wave_dir
                          << "\n";
                return 1;
            }
            std::sort(step_files[di].begin(), step_files[di].end(),
                      [](const RecordFile& a, const RecordFile& b) { return a.rank < b.rank; });
        }

        int n_ranks = (int)step_files[0].size();
        if (verbose)
            std::cout << "[wavefield2vtk_detail] Step " << step_num << " (" << n_ranks
                      << " rank(s))\n";

        // Per-direction vertex strain arrays [n_vert, 6]
        std::vector<float> dir_strain[3];
        for (int di = 0; di < 3; ++di)
            dir_strain[di].assign(n_vert * 6, 0.0f);

        // Read and scatter per direction, per rank
        for (int di = 0; di < 3; ++di) {
            for (int ri = 0; ri < n_ranks; ++ri) {
                int64_t n_local = (int64_t)vertex_id_list[ri].size();
                if (n_local == 0)
                    continue;
                auto sbuf = read_step_strain(step_files[di][ri].path, n_local);
                scatter_strain(sbuf, vertex_id_list[ri], n_vert, dir_strain[di]);
            }
        }

        // ── Write VTK ────────────────────────────────────────────────
        char out_path[256];
        std::snprintf(out_path, sizeof(out_path), "%s/wavefield_%d.vtk", vtk_dir.c_str(),
                      step_num);
        if (verbose)
            std::cout << "  Writing " << out_path << "\n";

        try {
            gf_vtk::VtkWriter vtk(out_path, "wavefield snapshot converted to VTK (C++ detail)");
            vtk.write_points(vtx_coords_f32);

            // Hex cells
            int64_t n_hex = n_cell;
            std::vector<int32_t> hex_cells(n_hex * 9);
            for (int64_t ci = 0; ci < n_hex; ++ci) {
                hex_cells[ci * 9] = 8;
                for (int j = 0; j < 8; ++j)
                    hex_cells[ci * 9 + 1 + j] = (int32_t)connectivity[ci * 8 + j];
            }
            std::vector<int32_t> hex_types(n_hex, 12);
            vtk.write_cells(hex_cells, hex_types);

            // Cell data: PML flag
            vtk.begin_cell_data(n_hex);
            std::vector<float> pml_f(n_hex);
            for (int64_t ei = 0; ei < n_hex; ++ei)
                pml_f[ei] = (float)is_pml[ei];
            vtk.write_scalar_field("PML_flag", pml_f);

            // Point data: 18 strain fields
            vtk.begin_point_data(n_vert);
            for (size_t fi = 0; fi < strain_field_names.size(); ++fi) {
                int di = (int)(fi / 6), ci = (int)(fi % 6);
                std::vector<float> pd(n_vert);
                for (int64_t vi = 0; vi < n_vert; ++vi)
                    pd[vi] = dir_strain[di][vi * 6 + ci];
                vtk.write_scalar_field(strain_field_names[fi], pd);
            }

        } catch (std::exception& e) {
            std::cerr << "  Error writing " << out_path << ": " << e.what() << "\n";
        }
    }

    if (verbose)
        std::cout << "  Done.\n";
    return 0;
}